apps/upgrade-journey/containers.go

```diff
@@ -0,0 +1,160 @@
+}
+
+func newCluster(nodeCount int) *cluster {
+	rootDir, err := os.Getwd()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return &cluster{
+		nodeCount:   nodeCount,
+		networkName: fmt.Sprintf("weaviate-upgrade-journey-%d", rand.Int()),
+		rootDir:     rootDir,
+		containers:  make([]testcontainers.Container, nodeCount),
+	}
+}
+
+func (c *cluster) startAllNodes(ctx context.Context, version string) error {
+	for i := 0; i < c.nodeCount; i++ {
+		container, err := c.startWeaviateNode(ctx, i, version)
+		if err != nil {
+			return err
+		}
+
+		c.containers[i] = container
+	}
+
+	return nil
+}
+
+func (c *cluster) rollingUpdate(ctx context.Context, version string) error {
+	log.Printf("starting rolling update to %s", version)
+	for i := 0; i < c.nodeCount; i++ {
+		if err := c.containers[i].Terminate(ctx); err != nil {
+			return err
+		}
+
+		// c.containers[i].StopLogProducer().Error
+
+		container, err := c.startWeaviateNode(ctx, i, version)
+		if err != nil {
+			log.Print(err)
+			logReader, logErr := container.Logs(context.Background())
+			if logErr != nil {
+				log.Fatal(logErr)
+			}
+
+			io.Copy(os.Stdout, logReader)
+			return err
+		}
+
+		c.containers[i] = container
+	}
+
+	log.Printf("completed rolling update to %s", version)
+	return nil
+}
+
+func (c *cluster) startNetwork(ctx context.Context) error {
+	_, err := testcontainers.GenericNetwork(ctx, testcontainers.GenericNetworkRequest{
+		NetworkRequest: testcontainers.NetworkRequest{
+			Name:     c.networkName,
+			Internal: false,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("network %s: %w", c.networkName, err)
+	}
+
+	return nil
+}
+
+func (c *cluster) volumePath(nodeId int) string {
+	return path.Join(c.rootDir, "data/", c.hostname(nodeId))
+}
+
+func (c *cluster) startWeaviateNode(ctx context.Context, nodeId int, version string) (testcontainers.Container, error) {
+	if err := os.MkdirAll(c.volumePath(nodeId), 0o777); err != nil {
+		log.Print(err)
+		return nil, err
+	}
+
+	image := fmt.Sprintf("semitechnologies/weaviate:%s", version)
+	container, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		Logger: log.Default(),
+		ContainerRequest: testcontainers.ContainerRequest{
+			Name:         fmt.Sprintf("%s-%d", c.hostname(nodeId), counter),
+			Image:        image,
+			Cmd:          []string{"--host", "0.0.0.0", "--port", "8080", "--scheme", "http"},
+			Networks:     []string{c.networkName},
+			ExposedPorts: []string{fmt.Sprintf("%d:8080", 8080+nodeId)},
+			AutoRemove:   false,
+			Env: map[string]string{
+				"QUERY_DEFAULTS_LIMIT":                    "25",
+				"AUTHENTICATION_ANONYMOUS_ACCESS_ENABLED": "true",
+				"PERSISTENCE_DATA_PATH":                   "/var/lib/weaviate",
+				"DEFAULT_VECTORIZER_MODULE":               "none",
+				"ENABLE_MODULES":                          "",
+				"CLUSTER_GOSSIP_BIND_PORT":                "7100",
+				"CLUSTER_DATA_BIND_PORT":                  "7101",
```

Comment:
do these ports need to be different on different nodes?

Reply:
No, since we're only accessing them from within the docker network. So `weaviate-0:7100` vs `weaviate-1:7100`, etc.